orbit/pkg/setup_experience/setup_experience.go

```diff
@@ -13,28 +22,265 @@ type Client interface {
 // SetupExperiencer is the type that manages the Fleet setup experience flow during macOS Setup
 // Assistant. It uses swiftDialog as a UI for showing the status of software installations and
 // script execution that are configured to run before the user has full access to the device.
+// If the setup experience is supposed to run, it will launch a single swiftDialog instance and then
+// update that instance based on the results from the /orbit/setup_experience/status endpoint.
 type SetupExperiencer struct {
 	OrbitClient Client
+	closeChan   chan struct{}
+	rootDirPath string
+	// Note: this object is not safe for concurrent use. Since the SetupExperiencer is a singleton,
+	// its Run method is called within a WaitGroup,
+	// and no other parts of Orbit need access to this field (or any other parts of the
+	// SetupExperiencer), it's OK to not protect this with a lock.
+	sd *swiftdialog.SwiftDialog
+	// Name of each step -> is that step done
+	steps   map[string]bool
+	started bool
 }
 
-func NewSetupExperiencer(client Client) *SetupExperiencer {
-	return &SetupExperiencer{OrbitClient: client}
+func NewSetupExperiencer(client Client, rootDirPath string) *SetupExperiencer {
+	return &SetupExperiencer{
+		OrbitClient: client,
+		closeChan:   make(chan struct{}),
+		steps:       make(map[string]bool),
+		rootDirPath: rootDirPath,
+	}
 }
 
 func (s *SetupExperiencer) Run(oc *fleet.OrbitConfig) error {
-	// We should only launch swiftDialog if we get the notification from Fleet.
 	if !oc.Notifications.RunSetupExperience {
 		log.Debug().Msg("skipping setup experience")
 		return nil
 	}
 
-	// poll the status endpoint
-	_, err := s.OrbitClient.GetSetupExperienceStatus()
+	_, binaryPath, _ := update.LocalTargetPaths(
+		s.rootDirPath,
+		"swiftDialog",
+		update.SwiftDialogMacOSTarget,
+	)
+
+	if _, err := os.Stat(binaryPath); err != nil {
+		return nil
+	}
+
+	// Poll the status endpoint. This also releases the device if we're done.
+	payload, err := s.OrbitClient.GetSetupExperienceStatus()
 	if err != nil {
 		return err
 	}
 
-	// TODO: fill this in!
+	// If swiftDialog isn't up yet, then launch it
+	if err := s.startSwiftDialog(binaryPath, payload.OrgLogoURL); err != nil {
+		return err
+	}
+
+	// Defer this so that s.started is only false the first time this function runs.
+	defer func() { s.started = true }()
+
+	select {
+	case <-s.closeChan:
+		log.Debug().Str("receiver", "setup_experiencer").Msg("swiftDialog closed")
+		return nil
+	default:
+		// ok
+	}
```

Comment:
@jahzielv just caught this while looking at something else.  Did you mean to block on the channel read?  unless the channel is already closed at this point, you'll lose the log and return

Reply:
also the reverse if `startSwiftDialog` doesn't send on the channel before this default, that routine will hang

Reply:
Hey yeah, this is meant to catch the channel write here: https://github.com/fleetdm/fleet/pull/22972/files#diff-3bc5f68723027852193160107b5df866f09f9a72b76aa881726668901b51e13cR259